File: cpp/src/generator.cpp
#include "generator.hpp"

namespace wildling {

Generator::Generator(std::string input_pattern, const Dictionaries& dictionaries)
    : source_(std::move(input_pattern)),
      tokens_(parse_pattern(source_, dictionaries)),
      count_(1) {
    for (const auto& token : tokens_) {
        count_ *= token.count();
    }
}

const std::string& Generator::source() const {
    return source_;
}

int Generator::count() const {
    return count_;
}

const std::vector<Token>& Generator::tokens() const {
    return tokens_;
}

std::string Generator::get(int index) const {
    if (index > count_ - 1 || index < 0) {
        return "";
    }
    std::string out;
    int index_with_offset = index;
    for (const auto& token : tokens_) {
        out += token.get(index_with_offset % token.count());
        index_with_offset /= token.count();
    }
    return out;
}

}  // namespace wildling

File: rust/src/generator.rs
use crate::parse_pattern::{parse_pattern, Dictionaries};
use crate::token::Token;

#[derive(Clone, Debug)]
pub struct Generator {
    pub source: String,
    tokens: Vec<Token>,
    count: i32,
}

impl Generator {
    pub fn new(input_pattern: &str, dictionaries: &Dictionaries) -> Self {
        let tokens = parse_pattern(input_pattern, dictionaries);
        let mut count = 1;
        for token in &tokens {
            count *= token.count();
        }
        Self {
            source: input_pattern.to_string(),
            tokens,
            count,
        }
    }

    pub fn count(&self) -> i32 {
        self.count
    }

    pub fn tokens(&self) -> &[Token] {
        &self.tokens
    }

    pub fn get(&self, index: i32) -> String {
        if index > self.count - 1 || index < 0 {
            return String::new();
        }
        let mut out = String::new();
        let mut index_with_offset = index;
        for token in &self.tokens {
            out.push_str(&token.get(index_with_offset % token.count()));
            index_with_offset /= token.count();
        }
        out
    }
}
